File: LeetCode/3. Longest Substring Without Repeating Characters/3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int best = 0, i = 0, n = s.size();
        map<char, int> seen;
        for (int j = 0; j < n; ++j) {
            // Add s[j]
            while (seen[s[j]] > 0 && i < j) {
                seen[s[i++]]--;
            }
            seen[s[j]]++;
            best = max(best, j - i + 1);
        }
        return best;
    }
};

File: LeetCode/3. Longest Substring Without Repeating Characters/3.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    best = 0
    i = 0
    seen = {}
    seen.default = 0
    # Possible solution is s[i, j)
    s.each_char.each_with_index do |_, j|
        while seen[s[j]] > 0 and i < j
            seen[s[i]] = seen[s[i]] - 1
            i += 1
        end
        seen[s[j]] += 1
        best = [best, j - i + 1].max
    end
    return best
end
